File: backend/src/main/java/org/luxons/sevenwonders/game/cards/Hands.java
package org.luxons.sevenwonders.game.cards;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.luxons.sevenwonders.game.api.HandCard;
import org.luxons.sevenwonders.game.api.Table;

public class Hands {

    private final int nbPlayers;

    private Map<Integer, List<Card>> hands;

    Hands(Map<Integer, List<Card>> hands, int nbPlayers) {
        this.hands = hands;
        this.nbPlayers = nbPlayers;
    }

    public List<Card> get(int playerIndex) {
        if (!hands.containsKey(playerIndex)) {
            throw new PlayerIndexOutOfBoundsException(playerIndex);
        }
        return hands.get(playerIndex);
    }

    public List<HandCard> createHand(Table table, int playerIndex) {
        return hands.get(playerIndex)
                    .stream()
                    .map(c -> new HandCard(c, table, playerIndex))
                    .collect(Collectors.toList());
    }

    public Hands rotate(HandRotationDirection direction) {
        Map<Integer, List<Card>> newHands = new HashMap<>(hands.size());
        for (int i = 0; i < nbPlayers; i++) {
            int newIndex = Math.floorMod(i + direction.getIndexOffset(), nbPlayers);
            newHands.put(newIndex, hands.get(i));
        }
        return new Hands(newHands, nbPlayers);
    }

    public boolean isEmpty() {
        return hands.values().stream().allMatch(List::isEmpty);
    }

    public boolean maxOneCardRemains() {
        return hands.values().stream().mapToInt(List::size).max().orElse(0) <= 1;
    }

    public List<Card> gatherAndClear() {
        List<Card> remainingCards = hands.values().stream().flatMap(List::stream).collect(Collectors.toList());
        hands.clear();
        return remainingCards;
    }

    class PlayerIndexOutOfBoundsException extends ArrayIndexOutOfBoundsException {
        PlayerIndexOutOfBoundsException(int index) {
            super(index);
        }
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/cards/Hands.kt
package org.luxons.sevenwonders.game.cards

import org.luxons.sevenwonders.game.api.HandCard
import org.luxons.sevenwonders.game.api.Table

class Hands internal constructor(private val hands: List<List<Card>>) {

    val isEmpty: Boolean = this.hands.all(List<Card>::isEmpty)

    operator fun get(playerIndex: Int): List<Card> {
        return hands[playerIndex]
    }

    fun discardHand(playerIndex: Int): Hands {
        val mutatedHands = hands.toMutableList()
        mutatedHands[playerIndex] = emptyList()
        return Hands(mutatedHands)
    }

    fun remove(playerIndex: Int, card: Card): Hands {
        val mutatedHands = hands.toMutableList()
        mutatedHands[playerIndex] = hands[playerIndex] - card
        return Hands(mutatedHands)
    }

    fun createHand(table: Table, playerIndex: Int): List<HandCard> {
        return hands[playerIndex].map { c -> HandCard(c, table, playerIndex) }
    }

    fun rotate(direction: HandRotationDirection): Hands {
        val newHands = when (direction) {
            HandRotationDirection.RIGHT -> hands.takeLast(1) + hands.dropLast(1)
            HandRotationDirection.LEFT -> hands.drop(1) + hands.take(1)
        }
        return Hands(newHands)
    }

    fun maxOneCardRemains(): Boolean = hands.map { it.size }.max() ?: 0 <= 1
}
